presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java

```diff
@@ -459,7 +459,9 @@ else if (symbolExpression instanceof Cast) {
 
             Type leftType = expressionTypes.get(comparison.getLeft());
             Type rightType = expressionTypes.get(comparison.getRight());
-            checkArgument(leftType.equals(rightType), "left and right type do not match in comparison expression (%s)", comparison);
+
+            // TODO: re-enable this check once we fix the type coercions in the optimizers
+            // checkArgument(leftType.equals(rightType), "left and right type do not match in comparison expression (%s)", comparison);
```

Comment:
Concerning correctness on the safe side it would be to return Optional.empty() here if types do not match.
But it can introduce performance degradation for some cases which used to work fast before
checkstate was introduced, but fail with exception now.
Apart from that it looks good.